### ISSUE TITLE
Pod crashlooping when an apiservice is unavailable

### ISSUE THREAD

**Issue by rust84** [ricoberger/vault-secrets-operator#22]
This was actually an issue with the cert-manager webhook api service but when it occurs it sends the secrets operator into a crashloop. Helm added some additional error handling in a recent [commit](https://github.com/johannges/helm/commit/f39e13b8dda6689746beb1967e3fa64ecfa0e95a) to handle when an api is unavailable. It might be preferable to apply a similar approach in the vault operator to avoid the pod crashing.

related to 

https://github.com/jetstack/cert-manager/issues/2273
https://github.com/helm/helm/issues/6361

Log

    {"level":"info","ts":1574070345.8219388,"logger":"cmd","msg":"Could not generate and serve custom resource metrics","error":"discovering resource information failed for VaultSecret in ricoberger.de/v1alpha1: unable to retrieve the complete list of server APIs: webhook.cert-manager.io/v1beta1: the server is currently unable to handle the request"}
    {"level":"info","ts":1574070349.1773176,"logger":"metrics","msg":"Metrics Service object updated","Service.Name":"vault-secrets-operator-metrics","Service.Namespace":"kube-system"}
    {"level":"info","ts":1574070352.479337,"logger":"cmd","msg":"Could not create ServiceMonitor object","error":"unable to retrieve the complete list of server APIs: webhook.cert-manager.io/v1beta1: the server is currently unable to handle the request"}

**Comment by ricoberger** [comment]
Hi @rust84 and thanks for reporting.

Do you have an example to reproduce the issue. As far as I can take it from the logs the operator should not be crashing, if he couldn't create the ServiceMonitor.

I'll take a closer look this week. If you have any further hints this would be very useful.

**Comment by rust84** [comment]
Had this happen again tonight. The metrics apiserver stopped responding an dvault operator failed to start. To reproduce this using the metrics-server you could possibly block port 443 between the nodes to make the metrics server fail.

Apiservice

        v1beta1.metrics.k8s.io                 kube-system/metrics-server   False (MissingEndpoints)   116m

Operator log

    {"level":"info","ts":1575236956.8585935,"logger":"cmd","msg":"Version: 1.4.0"}
    {"level":"info","ts":1575236956.858621,"logger":"cmd","msg":"Branch: HEAD"}
    {"level":"info","ts":1575236956.85863,"logger":"cmd","msg":"Revision: 7636ee6334edc4e0b9d1f08ffba015a55c1c876d"}
    {"level":"info","ts":1575236956.858637,"logger":"cmd","msg":"Go Version: go1.13.3"}
    {"level":"info","ts":1575236956.858644,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
    {"level":"info","ts":1575236956.8586512,"logger":"cmd","msg":"Version of operator-sdk: v0.11.0"}
    {"level":"info","ts":1575236956.858658,"logger":"cmd","msg":"Build User: runner"}
    {"level":"info","ts":1575236956.85866,"logger":"cmd","msg":"Build Date: 2019-10-28@18:12:43"}
    {"level":"info","ts":1575236956.858679,"logger":"vault","msg":"Reconciliation is enabled.","ReconciliationTime":300}
    {"level":"info","ts":1575236956.8910656,"logger":"vault","msg":"Renew Vault token"}
    {"level":"info","ts":1575236956.8925126,"logger":"leader","msg":"Trying to become the leader."}
    {"level":"info","ts":1575236960.3179264,"logger":"leader","msg":"Found existing lock with my name. I was likely restarted."}
    {"level":"info","ts":1575236960.317993,"logger":"leader","msg":"Continuing as the leader."}
    {"level":"info","ts":1575236960.3180003,"logger":"cmd","msg":"Watch all namespaces"}
    {"level":"info","ts":1575236963.7249851,"logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":"0.0.0.0:8383"}
    {"level":"info","ts":1575236963.7251725,"logger":"cmd","msg":"Registering Components."}
    {"level":"info","ts":1575236963.7254696,"logger":"controller-runtime.controller","msg":"Starting EventSource","controller":"vaultsecret-controller","source":"kind source: /, Kind="}
    {"level":"info","ts":1575236963.7255864,"logger":"controller-runtime.controller","msg":"Starting EventSource","controller":"vaultsecret-controller","source":"kind source: /, Kind="}
    {"level":"info","ts":1575236967.1279228,"logger":"cmd","msg":"Could not generate and serve custom resource metrics","error":"discovering resource information failed for VaultSecret in ricoberger.de/v1alpha1: unable to retrieve the complete list of server APIs: metrics.k8s.io/v1beta1: the server is currently unable to handle the request"}

**Comment by ricoberger** [comment]
Hi @rust84, thanks for the update. I was able to reproduce the issue now and try to fix it.

Steps to reproduce the issue:

- Install metrics-server (https://github.com/kubernetes-sigs/kind/issues/398):

```sh
kubectl apply -f https://gist.githubusercontent.com/hjacobs/69b6844ba8442fcbc2007da316499eb4/raw/5b8678ac5e11d6be45aa98ca40d17da70dcb974f/kind-metrics-server.yaml
```

- Install Prometheus Operator

```sh
helm upgrade --install --namespace monitoring -f tmp/prometheus-operator.yaml prometheus-operator stable/prometheus-operator
```

```yaml
defaultRules:
  create: false

global:
  rbac:
    create: true
    pspEnabled: true

alertmanager:
  enabled: false

grafana:
  enabled: false

coreDns:
  enabled: false

kubeDns:
  enabled: false

prometheus:
  enabled: true
  prometheusSpec:
    enableAdminAPI: true
    externalLabels:
      cluster: test
      replica: one
    retention: 30m
```

- Delete the `metrics-server` service:

```sh
k delete svc metrics-server --namespace kube-system
```

- Restart the Vault Secrets Operator

**Comment by ricoberger** [comment]
The issue should be fixed in version [1.4.2](https://github.com/ricoberger/vault-secrets-operator/releases/tag/1.4.2). If you have any problems feel free to reopen or create a new issue.

Thanks for pointing out the issue.

**Comment by rust84** [comment]
Thanks for the fix :+1:
